### ISSUE TITLE
spring cloud zuul is very slow,200% performance cut down after using zuul

### ISSUE THREAD

**Issue by ExtremeYu** [spring-cloud/spring-cloud-netflix#2247]
![a 56 36qf3qdm08xj1 kx](https://user-images.githubusercontent.com/16620268/29960379-f7e9a2c6-8f2d-11e7-9dc9-281caeab035f.png)

**Comment by ExtremeYu** [comment]
The faster is directly send request to the target service,the slower is request zuul gateway first ,then requst the target service, its really slow that zuul expend too many time to proxy the request,why?

**Comment by ExtremeYu** [comment]
![image](https://user-images.githubusercontent.com/16620268/29960764-8e1b1d96-8f2f-11e7-923d-49cdfc35326b.png)
BTW,I find zuul logged the **org.apache.http.wire**,i think zuul serialilize the json object and then deserialize ,its no use to do that




[网络] | Deserialize; d
-- | --

**Comment by gmarziou** [comment]
> BTW,I find zuul logged the org.apache.http.wire,i think zuul serialize the json object and then deserialize ,its no use to do that

It does it only if you set logging level to DEBUG which would be a bad idea for measuring performance.

Your comparison does not seem fair, have you tried to compare with another reverse proxy like nginx?

**Comment by spencergibb** [comment]
Please reopen if you find something else besides debug logging the apache http client.

**Comment by ExtremeYu** [comment]
i find a simple way,just change to okHttp3 and then the performance become normal,just  expend 5ms-6ms more ,the httpclient spend 20ms

**Comment by zywj** [comment]
I also encounter this problem when using the Finchley version. 
The zuul performance is half of the nginx.

**Comment by spencergibb** [comment]
@zywj have you done any of the suggestions above?

**Comment by ExtremeYu** [comment]
@zywj 老哥，换成okhttp,httpclient太垃圾

**Comment by ExtremeYu** [comment]
@spencergibb i suggest spring cloud should never use httpclient as default ,change to okhttp3 instead, best if support netty.

**Comment by zywj** [comment]
@ExtremeYu 我zuul后面有web服务，然后web服务调用api服务的。如果是zuul的httpclient的性能问题，那么在web服务请求api服务时，也会存在性能问题。但是我直接访问web服务对应的地址，性能是跟nginx做代理一样的，所以我还是觉得是zuul的问题。

**Comment by ExtremeYu** [comment]
zuul 本身就有很大性能损耗，我这里测试，加入zuul后，接口响应时间从平均8ms 增加到12ms，还是可以接受的，你先把全部工程换成okhttp再试

**Comment by zywj** [comment]
Yes. Zuul is CPU-bound, after read the article [Comparing API Gateway Performances: NGINX vs. ZUUL ](https://engineering.opsgenie.com/comparing-api-gateway-performances-nginx-vs-zuul-vs-spring-cloud-gateway-vs-linkerd-b2cc59c65369) and [THE RIOT GAMES API: FULFILLING ZUUL'S DESTINY](https://engineering.riotgames.com/news/riot-games-api-fulfilling-zuuls-destiny).

So. @ExtremeYu . You may be right. The `httpclient` may limits the Zuul performance.

**Comment by jieblog** [comment]
吓得我也去换成okttp试试

**Comment by zywj** [comment]
@jieblog 如果ZUUL的CPU使用率没有到100%的话，ZUUL性能还是接近nginx的。

**Comment by jieblog** [comment]
常年 100%状态

**Comment by OuIChien** [comment]
大家遇到过，压测时cpu使用率间歇性下降到个位数吗？然后服务就失去响应了（不是超时）。

**Comment by masykurm** [comment]
Hi all..
I also experiencing this issue where zuul at some point giving me high latency above 1 second, while the downstream service in which zuul connected to only took arouns 6-10 ms

Try using okhttp still didn't help.. here is what I set in properties

zuul.ribbon.okhttp.enabled=true
ribbon.okhttp.enabled=true

Does my configuration already correct ?
I'm not using eureka for service discovery, I'm using static list of ip addresses

I am using 
Spring Boot 1.5.9 release
Spring Cloud version Dalston Release

Thank you

**Comment by ryanjbaxter** [comment]
Try using the latest edgware release, if you are still having issues please open a separate issue with a sample and a way to reproduce the problem.

**Comment by masykurm** [comment]
hi @ryanjbaxter ,
 
thanks for the warm feedback.. I will try edgware release, and will let you know the results.
Additionally, is there any specific configuration I need to perform in this version ?

**Comment by masykurm** [comment]
Some updates on using edgware release
1. response time is getting better than previous one , although : 
            a. 99.5th percentile : 20x higher
            b. 99th percentile : 10x higher
            c. 50th percentile : almost the same
    (using prometheus metrics)

2. some configuration I have made:
      a. disable all unnecessary built-in zuul filter
      b. increase heap size and core size of the instance
      c. override SendResponseFilter with custom filter
      d. since I'm using Undertow, tune a bit on the buffer size, io threads, and worker threads
      e. using OkHttp client instead of Http Client

Still finding out another way to tune the 99th-ish percentile response time. 
Not sure if the 99th-ish metrics incorrectly calculated or it is really the response time, but during  observing the duration in log file is process is completed  on around 50th percentile response time
 
@ryanjbaxter  - most welcome for next / further advice

Thank you, appreciate for your response / feedback

**Comment by ryanjbaxter** [comment]
The only thing I can suggest is digging deeper to see if you can identify underlying spring cloud netflix code that could be causing an issue.  If you can point us towards a problem we can help address it  There is a lot of different components involved here and nothing you have told us so far indicates a problem with Spring Cloud Netflix.

**Comment by masykurm** [comment]
Hi @ryanjbaxter ,

Thanks for your feedback. Related with Spring Cloud Netflix itself, I have done following things and deep dive bit on the Zuul part

1.  Disable `SendPostFilter`, `DebugFilter`, `FormBodyWrapperFilter` . 
    This is after I found out that `FilterProcessor` class will loop through all enabled filter to be executed. However I guess, it will still got loaded in the list of filters, since conditional checking on `disabled` flag performed on `filter.runFilter()` in `FilterProcessor` class. Maybe can do pre-check in `FilterLoader.getFiltersByType` method  in parallel with loading all available filters so that only enabled filters are loaded

2.  Create custom post Filter that extends `SendPostFilter` class and add following code (closing response data stream)

    ```
    @Override
    public Object run() {
    	Object result = null;
        try {
             result = super.run();
             // Simply force a close for every request...
             doCloseSourceInputStream();
        } catch (Exception ex) {
        	// Simply force a close for every request...
            doCloseSourceInputStream();
            ReflectionUtils.rethrowRuntimeException(ex);
        }
        return result;
    }

    private void doCloseSourceInputStream() {
        RequestContext context = RequestContext.getCurrentContext();
        InputStream inputStream = context.getResponseDataStream();
        if (Objects.nonNull(inputStream)) {
            try {
                inputStream.close();
            } catch (IOException e) {
                // This should actually never happen
                log.error("Failed to close wrapped stream", e);
            }
        }
    }
    ```

**Comment by ryanjbaxter** [comment]
So I am still looking for where in spring cloud netflix you see the problem

**Comment by masykurm** [comment]
Hi @ryanjbaxter 
Sorry for the long wait, I just figure out to obtain the error log. After observing all the traffic logs, I got following errors. Somehow I got `UndeclaredThrowableException` in which raised by `SendResponseFilter`.`run` on line 120

```
org.springframework.util.ReflectionUtils.rethrowRuntimeException [317]
org.springframework.cloud.netflix.zuul.filters.post.SendResponseFilter.run [120]
mine.proxy.filter.post.InputStreamClosingSendResponsePostFilter.run [50]
com.netflix.zuul.ZuulFilter.runFilter [112]
com.netflix.zuul.FilterProcessor.processZuulFilter [193]
com.netflix.zuul.FilterProcessor.runFilters [157]
com.netflix.zuul.FilterProcessor.postRoute [92]
com.netflix.zuul.ZuulRunner.postRoute [87]
com.netflix.zuul.http.ZuulServlet.postRoute [107]
com.netflix.zuul.http.ZuulServlet.service [88]
org.springframework.web.servlet.mvc.ServletWrappingController.handleRequestInternal [157]
org.springframework.cloud.netflix.zuul.web.ZuulController.handleRequest [44]
org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle [50]
org.springframework.web.servlet.DispatcherServlet.doDispatch [967]
org.springframework.web.servlet.DispatcherServlet.doService [901]
org.springframework.web.servlet.FrameworkServlet.processRequest [970]
org.springframework.web.servlet.FrameworkServlet.doPost [872]
javax.servlet.http.HttpServlet.service [707]
org.springframework.web.servlet.FrameworkServlet.service [846]
javax.servlet.http.HttpServlet.service [790]
io.undertow.servlet.handlers.ServletHandler.handleRequest [85]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [129]
org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal [55]
org.springframework.web.filter.OncePerRequestFilter.doFilter [107]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal [110]
org.springframework.web.filter.OncePerRequestFilter.doFilter [107]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
org.springframework.web.filter.RequestContextFilter.doFilterInternal [99]
org.springframework.web.filter.OncePerRequestFilter.doFilter [107]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal [108]
org.springframework.web.filter.OncePerRequestFilter.doFilter [107]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal [81]
org.springframework.web.filter.OncePerRequestFilter.doFilter [107]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
org.springframework.cloud.sleuth.instrument.web.TraceFilter.doFilter [166]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal [197]
org.springframework.web.filter.OncePerRequestFilter.doFilter [107]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal [106]
org.springframework.web.filter.OncePerRequestFilter.doFilter [107]
io.undertow.servlet.core.ManagedFilter.doFilter [61]
io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter [131]
io.undertow.servlet.handlers.FilterHandler.handleRequest [84]
io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest [62]
io.undertow.servlet.handlers.ServletChain$1.handleRequest [64]
io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest [36]
io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest [131]
io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest [57]
io.undertow.server.handlers.PredicateHandler.handleRequest [43]
io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest [46]
io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest [64]
io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest [60]
io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest [77]
io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest [43]
io.undertow.server.handlers.PredicateHandler.handleRequest [43]
io.undertow.server.handlers.PredicateHandler.handleRequest [43]
io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest [292]
io.undertow.servlet.handlers.ServletInitialHandler.access$100 [81]
io.undertow.servlet.handlers.ServletInitialHandler$2.call [138]
io.undertow.servlet.handlers.ServletInitialHandler$2.call [135]
io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call [48]
io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call [43]
io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest [272]
io.undertow.servlet.handlers.ServletInitialHandler.access$000 [81]
io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest [104]
io.undertow.server.Connectors.executeRootHandler [332]
io.undertow.server.HttpServerExchange$1.run [830]
java.util.concurrent.ThreadPoolExecutor.runWorker [1149]
java.util.concurrent.ThreadPoolExecutor$Worker.run [624]
java.lang.Thread.run [748]
```
In `SendResponseFilter.run` line 120, I'm calling `super.run()` in which invokes method `SendResponseFilter.run()` as in previous post

After I look the code in the `SendResponseFilter` line 120, leads me on following code in which the exception thrown either by `addResponseHeaders` or `writeResponse`method 

```
@Override
public Object run() {
	try {
		addResponseHeaders();
		writeResponse();
	}
	catch (Exception ex) {
		ReflectionUtils.rethrowRuntimeException(ex);
	}
	return null;
}
```
Cannot figure out what really the cause 
Kindly need your advice

Thanks in advance

**Comment by ryanjbaxter** [comment]
Looks like part of the stacktrace is missing.

Is this your filter? 
`mine.proxy.filter.post.InputStreamClosingSendResponsePostFilter`

**Comment by masykurm** [comment]
Hi.. yes that is part of my filter that extend `SendResponseFilter`.

that is all the complete stacktrace from `ex.getStackTrace` method

**Comment by ryanjbaxter** [comment]
I cant really be much assistance if the exception is originating from your own filter.

At this point it would be useful if you could put together a project that reproduces the issue along with instructions on how to reproduce the problem and open a separate issue with that information.  This does not seem related to the original issue.

**Comment by masykurm** [comment]
Hi..
Thanks for the prompt response. Yes I agree, I guess it is separate issue. I found this when I observing the latency. Maybe I will disable my filter for time being..  and will let you know if I'm still encountering the same issue and open separate issue if any

Really appreciate for all your help and response

**Comment by masykurm** [comment]
Just to update that I'm still encountering same error after removing my custom filter. In http response wise, it give me status code `500` with error message `GENERAL`

It occurs intermittenly.

**Comment by vijaychd** [comment]
i am using Edgware.SR4 and noticing similar behavior.  zuul proxy response time is 1sec.  direct api call is 10-20 ms.
Any help is appreciated

**Comment by ryanjbaxter** [comment]
We can expect it to be slower going through the proxy, if you can identify a particular area of the code that is causing any kind of performance issue please open a separate issue with the details and how to reproduce it.

**Comment by dwivedis** [comment]
I am using **Spring Cloud Gateway** instead of **Zuul** to improve the (TPS)performance of the application but still we are getting less performance. I am sharing Gateway configuration below:

`spring:
  application:
    name: MyCloudGateway
  jmx:
    default-domain: MyCloudGateway
  cloud:
    gateway:
      routes:
      - id: SampleMicroService
        uri: http://127.0.0.1:9086
        predicates:
        - Path=/client1/**
        filters:
        - name: RequestRateLimiter
          args:
            redis-rate-limiter.replenishRate: 20000 
            redis-rate-limiter.burstCapacity: 40000  
            key-resolver: "#{@remoteAddrKeyResolver}"
        - StripPrefix=1    
        #- name: Hystrix
         # args:
          # name: SampleMicroService
           #fallbackUri: forward:/fallback 
      httpclient:
        pool:
          acquire-timeout: 6000
          max-connections: 10000
        connect-timeout: 60000
                   
  redis:
    host: 127.0.0.1
    port: 6379
server:
  port: 9087
  compression:
    enabled: false
    min-response-size: 1000
  connection-timeout: 50s
  max-http-header-size: 200000
#ribbon:
  #okhttp:
    #enabled: true
  #http:
    #client:
      #enabled: true    
  #tomcat:
   # accesslog:
    #  enabled: true
    #max-threads: 1000
    #min-spare-threads: 200
    
  #tomcat.uri-encoding: UTF-8
#eureka:
  #instance:
    #prefer-ip-address: true 
    #instance-id: ${spring.application.name}:${spring.application.instance_id:${random.value}}
hystrix:
 command:
  SampleMicroService:
   execution:
    isolation:
     thread:
      timeoutInMilliseconds:  120000   
   circuitBreaker:
        requestVolumeThreshold: 50000
        sleepWindowInMilliseconds: 60000
        errorThresholdPercentage: 100       
      
management:
  endpoints:
    web:
      exposure:
       include: hystrix.stream,turbine.stream 
               `
**Please suggest, where we are doing wrong.**

**Comment by spencergibb** [comment]
@dwivedis what do you think is wrong?

**Comment by dwivedis** [comment]
@spencergibb, Is any changes to do in configuration file to achieve 500 tps for normal string?

**Comment by spencergibb** [comment]
@dwivedis on what hardware with what test driver? There's a lot of ifs?

I get over 3k requests per second on my laptop https://github.com/spencergibb/spring-cloud-gateway-bench
